### PR TITLE
[ReshapeMapTool] Fix snapping matches with tracing

### LINKED ISSUE #57225: Repeated Crash when using Reshape Feature Tool
### What is the bug or the crash?

I'm working on digitizing historical political borders based on old maps.  Regularly (at least 5-10% of the time), when I right-click to complete a polygon reshape using the Reshape Feature tool, QGIS crashes (see stack trace below).

In other QGIS work that I've done, I've never encountered this kind of instability.  With a fresh install of QGIS and a fresh profile, I'm seeing the same behavior.

It doesn't appear to matter how complicated my polygons are, as I've had it crash when adding just a few vertices, and I've also had it work when adding thousands.  If it crashes (apart from one case, where it happened twice in a row) it always works after I re-open QGIS.

I'm not sure if it's relevant, but I'm using the snap feature to trace some of the existing geometries.

### Steps to reproduce the issue

1. Create a new layer (if it matters, I'm using a GeoPackage Layer of geometry MultiPolygon).
2. Create a polygon in the layer.
3. Using the "Reshape Features" tool from the Advanced Digitizing toolbar, reshape the polygon to add new area.
 a. If it matters, use the snap tool and tracing to trace at least part of some existing geometry as part of the reshape process.
4. Right click to complete, and observe QGIS crashing.
 a. (Happens about 5-10% of the time)

### Versions

QGIS Version: 3.36.1-Maidenhead
QGIS code revision: 3e589453
Compiled against Qt: 5.15.3
Running against Qt: 5.15.3
Compiled against GDAL: 3.8.4
Running against GDAL: 3.8.4

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

```Python Stack Trace
Windows fatal exception: access violation

Current thread 0x000015b8 (most recent call first):



Stack Trace


QgsPointXY::QgsPointXY :
QgsVectorLayerEditUtils::addTopologicalPoints :
QgsVectorLayer::addTopologicalPoints :
QgsMapToolReshape::reshape :
QgsMapToolReshape::cadCanvasReleaseEvent :
QgsMapToolAdvancedDigitizing::canvasReleaseEvent :
QgsMapCanvas::mouseReleaseEvent :
QWidget::event :
QFrame::event :
QGraphicsView::viewportEvent :
QCoreApplicationPrivate::sendThroughObjectEventFilters :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main :
BaseThreadInitThunk :
RtlUserThreadStart :




QGIS Info
QGIS Version: 3.36.1-Maidenhead
QGIS code revision: 3e589453
Compiled against Qt: 5.15.3
Running against Qt: 5.15.3
Compiled against GDAL: 3.8.4
Running against GDAL: 3.8.4



System Info
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.19045
```

### PR DESCRIPTION
**Pull request by troopa81**
`addCurve()` method already update the `mSnappingMatches ` variable so we need to update `pointBefore` variable after the `addCurve()` call

Fixes #57225

### CODE REVIEW

File: src/gui/maptools/qgsmaptoolcapture.cpp [hunk 1]
```diff
@@ -316,8 +316,8 @@ bool QgsMapToolCapture::tracingAddVertex( const QgsPointXY &point )
   mSnappingMatches.removeLast();
   mSnappingMatches.append( QgsPointLocator::Match() );
 
-  int pointBefore = mCaptureCurve.numPoints();
   addCurve( new QgsLineString( mapPoints ) );
+  int pointBefore = mCaptureCurve.numPoints();
```
Comment (nyalldawson) [inline src/gui/maptools/qgsmaptoolcapture.cpp:320]: Just to make it clearer what this is actually doing, could we move all the logic relating to pointBefore/pointAfter and the sync snapping matches into the block which starts with `if ( QgsSettingsRegistryCore::settingsDigitizingConvertToCurve->value() )` ? As far as I can see this logic ONLY applies now if the curve approximation is happening.

And on that note... if the unsegmentized line ends up with LESS vertices then the segmentized one, won't `mSnappingMatches` have too many entries and need to be trimmed? Maybe this whole logic could be simplified down to just:

```
    // sync the snapping matches list
    mSnappingMatches.resize( mCaptureCurve.numPoints() )
```

Comment (troopa81) [reply src/gui/maptools/qgsmaptoolcapture.cpp]: Good catch! Done c3c3ccc8812

File: src/gui/maptools/qgsmaptoolcapture.cpp [hunk 2]
```diff
@@ -343,12 +342,13 @@ bool QgsMapToolCapture::tracingAddVertex( const QgsPointXY &point )
...
       }
     }
-  }
 
-  // sync the snapping matches list
-  const int pointAfter = mCaptureCurve.numPoints();
-  for ( ; pointBefore < pointAfter; ++pointBefore )
-    mSnappingMatches.append( QgsPointLocator::Match() );
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+    mSnappingMatches = QVector( mCaptureCurve.numPoints(), QgsPointLocator::Match() ).toList();
```
Comment (nyalldawson) [inline src/gui/maptools/qgsmaptoolcapture.cpp:347]: This doesn't look right, we should be keeping the existing entries but resizing

Comment (troopa81) [reply src/gui/maptools/qgsmaptoolcapture.cpp]: done 54803792f9b

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**